### ISSUE TITLE
Move the handling of focusNextCell action from reducer to epics

### ISSUE THREAD

**Issue by bryanchen-d** [nteract/archived-desktop-app#5165]
**Is your feature request related to a problem? Please describe.**
I am implementing a read-only mode on top of the @nteract/core package, in which I am trying to disable adding new cells. 
Currently when user hit shift+enter on the last cell, it emits the focusNextCell and focusNextCellEditor actions, the logic of determining whether to insert a new cell is done in the reducer. Even though there is a `createCellIfUndefined` in focusNextCell action can be to false to skip creating a new cell, the focusNextCellEditor does not detect that and just assume the focus has been moved.

**Describe the solution you'd like**
I think instead of handling these logic in the reducer, it would be better to move them to epics which is more flexible for re-use. Reducer can handle actions closer to the redux store, in this case, I'd imagine 
- An epic to take care of the focusNextCell() action, determine whether or not a new cell needed to be created, if yes, emit a series of actions: createNewCell(cellId), setFocusCell(cellId), setFocusCellEditor(cellId)
- Reducer to handle createNewCell(), put the new cell into the cellMap
- Reducer to handle setFocusCell and setFocusCellEditor, update the id in the notebook model

**Describe alternatives you've considered**
An alternative solution would be to add logic in the determination of whether or not to emit the FOCUS_NEXT_CELL and FOCUS_NEXT_CELL_EDITOR at where they could be triggered, but first of all that logic is redundant since it is also done in the reducer, secondly this means every time a new place which can trigger these actions will have to do the same check, so better to centralize the check in one place.

@nteract/core

**Comment by captainsafia** [comment]
The overall scenario seems sensible. Although I don't think we need to move the logic to epics. We can support all the requirements in a reducer. It'll likely involve some changes to the actions.

>  Even though there is a createCellIfUndefined in focusNextCell action can be to false to skip creating a new cell, the focusNextCellEditor does not detect that and just assume the focus has been moved.

This sounds like a bug. We should probably add logic in our `focusNextCellEditor` action that checks to see if the focus is currently on the last cell.

That aside, we can probably do this entirely in reducers. All the focus and create actions end up modifying the state so we can cut the middleman and do it there.

We can create a new action `FOCUS_NEXT_CELL_AND_EDITOR` that looks like this:

```
{
  action: FOCUS_NEXT_CELL_AND_EDITOR,
  payload: {
    id: CellId;
  },
  meta: {
    createNewCell: boolean;
  }
}
```

Then in the reducer, we can do the following:

```
case FOCUS_NEXT_CELL_AND_EDITOR:
   if (noNextCell) {
     if (action.meta.createNewCell) {
        createNewCell(cellId);
     }
        setFocusCell(cellId);
        setFocusCellEditor(cellId);
    }
  }
```

We already have reducers for creating new cells and changing the focused cell.

**Comment by bryanchen-d** [comment]
I agree this would also work. The logic can exist in either reducers or epics, but I think what I am trying to say is moving logic to epics would make the design on epics and reducer more clear - that IMHO reducer only take care of simply storing data without extra calculation, and move all logic to epics. What do you think?

**Comment by captainsafia** [comment]
> that IMHO reducer only take care of simply storing data without extra calculation

The guidance is that reducers should be purely modifying the state. However, in practice, this ends up being a little hard to do. In fact, we have a lot of reducers in nteract that modify the state conditionally. In particular, I'm thinking of the reducers for handling display updates.

In general, I think that it's fine to have additional logic in the reducer as long as the logic is deterministic.

Tagging this for the July release.

**Comment by emeeks** [comment]
I wouldn't create a new `FOCUS_NEXT_CELL_AND_EDITOR` rather we should ensure that `FOCUS_NEXT_CELL_EDITOR` also focuses the cell. We shouldn't ever have a situation where the editor focus does not match the cell focus except in the case where the editor is not focused and only the cell is focused.

**Comment by captainsafia** [comment]
@emeeks That sensible too. Would you be interested in contributing the necessary changes here?
